Docs: add description of Indirect Condition problem (#4295)
ArchipelagoMW/Archipelago: docs/apworld_dev_faq.md

```py
item_pool += [self.create_filler() for _ in range(total_locations - len(item_pool))]
```

---

### I learned about indirect conditions in the world API document, but I want to know more. What are they and why are they necessary?

The world API document mentions how to use `multiworld.register_indirect_condition` to register indirect conditions and **when** you should use them, but not *how* they work and *why* they are necessary. This is because the explanation is quite complicated.

> Region sweep (the algorithm that determines which regions are reachable) is a Breadth-First Search of the region graph from the origin region, checking entrances one by one and adding newly reached nodes (regions) and their entrances to the queue until there is nothing more to check.

Reviewer: ```suggestion
Region sweep (the algorithm that determines which regions are reachable) is a Breadth-First Search of the region graph. It starts from the origin region, checks entrances one by one, and adds newly reached regions and their entrances to the queue until there is nothing more to check.
```

> For performance reasons, AP only checks every entrance once. However, if any entrance's access_rule depends on region access, then it is possible for this to happen:
> 1. An entrance that depends on a region is checked and determined to be nontraversable because the region hasn't been reached yet during the graph search.
> 2. After that, the region is reached by the graph search.

Reviewer: ```suggestion
For performance reasons, AP only checks every entrance once. However, if an entrance's access_rule depends on region access, then the following may happen:
1. The entrance is checked and determined to be nontraversable because the region in its access_rule hasn't been reached yet during the graph search.
2. Then, the region in its access_rule is determined to be reachable.
```

Just a bit more concise

> The entrance *would* now be determined to be traversable if it were rechecked, but it is not.

Reviewer: ```suggestion
This entrance *would* be called traversable if it were rechecked, but it won't be rechecked this cycle.
```

Author: "called traversable"? that doesn't seem like wording we use anywhere else maybe "marked traversable"? not sure
actually, maybe just "reachable"

Reviewer: Yeahh, maybe just reachable would be fine. idk, reachable sounds weird to me, but it's probably okay

> This entrance *would* be called reachable if it were rechecked, but it won't be rechecked this cycle.

Author: ```suggestion
This entrance *would* be in logic if it were rechecked, but it won't be rechecked this cycle.
```
slight rewording scipio and I agreed on in DMs

To account for this case, AP would have to recheck all entrances every time a new region is reached until no new regions are reached.

> However, there is a way to **manually** define that a *specific* entrance needs to be rechecked during region sweep if a *specific* region is reached during it. This is what an indirect condition is.

Reviewer: ```suggestion
An indirect condition is how you can **manually** define that a *specific* entrance needs to be rechecked during region sweep if a *specific* region is reached during it.
```
and idk about the \*\*, up to you

Author: all italics/bold formatting was written by vi and i have am ambivalent to, if you think it makes it read worse suggest them removed and I will

> This keeps almost all of the performance upsides. Even a game making heavy use of indirect conditions (See: The Witness) is still significantly faster than if it just blanket "rechecked all entrances until nothing new is found".

Reviewer: ```suggestion
This keeps most of the performance upsides. Even in a game making heavy use of indirect conditions (ex: The Witness), using them is significantly faster than just "rechecking each entrance until nothing new is found".
```

> The reason entrance access rules using `location.can_reach` and `entrance.can_reach` are also affected is simple: They call `region.can_reach` on their respective parent/source region.

Reviewer: ```suggestion
The reason entrance access rules using `location.can_reach` and `entrance.can_reach` are also affected is because they call `region.can_reach` on their respective parent/source region.
```

We recognize it can feel like a trap since it will not alert you when you are missing an indirect condition, and that some games have very complex access rules.

> As of [PR #3682 (Core: Region handling customization)](https://github.com/ArchipelagoMW/Archipelago/pull/3682) being merged, it is also possible for a world to opt out of indirect conditions entirely, although it does come at a flat performance cost.

Reviewer: ```suggestion
As of [PR #3682 (Core: Region handling customization)](https://github.com/ArchipelagoMW/Archipelago/pull/3682) being merged, it is possible for a world to opt out of indirect conditions entirely, instead using the system of checking each entrance whenever a region has been reached, although this does come with a performance cost.
```

> It should only be used by games that *really* need it. For most games, it should be reasonable to know all entrance &rarr; region dependencies, and in this case, indirect conditions are still preferred because they are faster.

Reviewer: ```suggestion
Opting out of using indirect conditions should only be used by games that *really* need it. For most games, it should be reasonable to know all entrance &rarr; region dependencies, making indirect conditions preferred because they are much faster.
```
